host/xtest/pkcs11_1000.c

```diff
@@ -3741,3 +3741,428 @@ static void xtest_pkcs11_test_1016(ADBG_Case_t *c)
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB_ENCRYPT_DATA,
+			    datalen, &derv_key_hdl, keylen, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_DATA_LEN_RANGE, rv))
+		goto out;
+
+	/* Data is not multiple of 16 */
+	datalen = 18;
+	keylen = 32;
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB_ENCRYPT_DATA,
+			    datalen, &derv_key_hdl, keylen, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_DATA_LEN_RANGE, rv))
+		goto out;
+
+	/* Wrong Mechanism */
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB,
+			    datalen, &derv_key_hdl, keylen, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_MECHANISM_INVALID, rv))
+		goto out;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	Do_ADBG_BeginSubCase(c, "Failure if operation already active");
+
+	/* Generate an AES key which can perform Encryption */
+	rv = C_GenerateKey(session, &cktest_aes_keygen_mechanism,
+			   parent_template_w_enc_der,
+			   ARRAY_SIZE(parent_template_w_enc_der),
+			   &aes_key_enc);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_EncryptInit(session, &cktest_aes_cbc_mechanism, aes_key_enc);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	datalen = 32;
+	keylen = 32;
+	rv = derive_sym_key(session, aes_key2, CKM_AES_ECB_ENCRYPT_DATA,
+			    datalen, &derv_key_hdl, keylen, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_OPERATION_ACTIVE, rv))
+		goto out;
+
+	rv = C_EncryptFinal(session, NULL, NULL);
+	/* Only check that the operation is no more active */
+	if (!ADBG_EXPECT_TRUE(c, rv != CKR_BUFFER_TOO_SMALL))
+		goto out;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	Do_ADBG_BeginSubCase(c, "Failure if parent key C_ENCRYPT is TRUE");
+
+	datalen = 32;
+	keylen = 32;
+	rv = derive_sym_key(session, aes_key_enc, CKM_AES_ECB_ENCRYPT_DATA,
+			    datalen, &derv_key_hdl, keylen, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_FUNCTION_FAILED, rv))
+		goto out;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	Do_ADBG_BeginSubCase(c, "Failure if parent key C_DERIVE is FALSE");
+
+	rv = C_DestroyObject(session, aes_key1);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_GenerateKey(session, &cktest_aes_keygen_mechanism,
+			   parent_template_wo_derive,
+			   ARRAY_SIZE(parent_template_wo_derive),
+			   &aes_key1);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	datalen = 32;
+	keylen = 32;
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB_ENCRYPT_DATA,
+			    datalen, &derv_key_hdl, keylen, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_KEY_FUNCTION_NOT_PERMITTED, rv))
+		goto out;
+
+out:
+	Do_ADBG_EndSubCase(c, NULL);
+
+close_session:
+	ADBG_EXPECT_CK_OK(c, C_CloseSession(session));
+
+close_lib:
+	ADBG_EXPECT_CK_OK(c, close_lib());
+}
+ADBG_CASE_DEFINE(pkcs11, 1017, xtest_pkcs11_test_1017,
+		 "PKCS11: Key Derivation tests");
```

Comment:
This starts to be quite long and I suppose other key derivation things also has some length so shall we just say here that it is AES key derivation tests?

Reply:
Sure, I will change this